### PR TITLE
Refactor fixed-form

### PR DESCRIPTION
**Pull request by certik**

### CODE REVIEW

**Comment by certik** [comment]
@konradha I think you'll like this PR: it simplifies the fixed-form tokenizer quite a bit. More work is needed.

File: tests/reference/ast-fixed_form_3-7f8625a.stderr [hunk 1]
```diff
@@ -1,9 +1,5 @@
...
- --> tests/errors/fixed_form_3.f:4:21 - 5:7
+ --> tests/errors/fixed_form_3.f:4:21
   |
-4 |    500   print *, sin(y
-  |                        ...
-...
-  |
-5 |          print *, "OK"
-  | ...^^^^^^^ 
+4 | 500   print *, sin(y
```
Comment (konradha) [inline tests/reference/ast-fixed_form_3-7f8625a.stderr:4]: Great, you fixed the location here.

File: src/lfortran/parser/fixedform_tokenizer.cpp [hunk 2]
```diff
@@ -444,17 +444,21 @@ struct FixedFormRecursiveDescent {
         locations.push_back(loc);
     }
 
-    // Same as push_token(), but update `cur` and `t.cur`
+    // Same as push_token_no_advance(), but update `cur` and `t.cur`
     // TODO: just use `token_type`, and advance `cur` accordingly and check
     // that the string actually matches
-    // TODO: add push_token_no_advance() which does not modify cur
     void push_token2(unsigned char *&cur, const std::string &token_str,
```
Comment (konradha) [inline src/lfortran/parser/fixedform_tokenizer.cpp:450]:
```suggestion
    void push_token_advance(unsigned char *&cur, const std::string &token_str,
```

I know you're doing this for the sake of not getting confused with overloading; for readability's sake I'd suggest using a precise meaning for the function(s):

File: src/lfortran/parser/fixedform_tokenizer.cpp [hunk 3]
```diff
@@ -444,17 +444,21 @@ struct FixedFormRecursiveDescent {
...
             yytokentype token_type) {
-        push_token(cur, token_str, token_type);
+        push_token_no_advance(cur, token_str, token_type);
         cur += token_str.size();
         t.cur = cur;
     }
 
+    // Same as push_token2(), but token_type is automatically determined
+    void push_token3(unsigned char *&cur, const std::string &token_str) {
```
Comment (konradha) [inline src/lfortran/parser/fixedform_tokenizer.cpp:458]:
```suggestion
    void push_token_advance_with_tokentype(unsigned char *&cur, const std::string &token_str) {
```

Same as above.

**Review by konradha — APPROVED** [review]
This is coming along really well, thanks! I think I'll make a pass adding some comments and maybe thinning it out some more (e.g. the global `lines` shouldn't be needed anymore I think, it was for bookkeeping purposes).

**Comment by certik** [comment]
Yes, I was going to do the rename later, so the number 2 and 3 was just a place holder. Yes, I don't like overloading, as it is confusing. However, I was going to simplify this further first, I suspect we might only need two versions, one with advance, the other one is no advance. I am doing it iteratively, and running tests at each change, as that's the only way I was able to simplify it. Once I am done in future PRs, I suspect it will be nice and clean, and then we can rename these names.

**Pull request merged**